Orm/Xtensive.Orm.Tests/Linq/ContainsOverKeysCollectionTest.cs

```diff
@@ -65,139 +98,593 @@ protected override void PopulateData()
+            .Where(e => new {
+              Key = e.Key,
+              EntityKey = e.Entity.Key,
+              DoubleEntityKey = e.Entity.Entity.Key
+            }.DoubleEntityKey.In(keyListAsCollection)).ToList());
+
+        Assert.Throws<QueryTranslationException>(() =>
+          result = session.Query.All<MyEntity3>()
+            .Where(e => new {
+              Key = e.Key,
+              EntityKey = e.Entity.Key,
+              DoubleEntityKey = e.Entity.Entity.Key
+           }.DoubleEntityKey.In(keyArray)).ToList());
+
+        Assert.Throws<QueryTranslationException>(() =>
+          result = session.Query.All<MyEntity3>()
+            .Where(e => new {
+              Key = e.Key,
+               EntityKey = e.Entity.Key,
+              DoubleEntityKey = e.Entity.Entity.Key
+           }.DoubleEntityKey.In(keyArrayAsEnumerable)).ToList());
+
+        Assert.Throws<QueryTranslationException>(() =>
+          result = session.Query.All<MyEntity3>()
+            .Where(e => new {
+              Key = e.Key,
+              EntityKey = e.Entity.Key,
+              DoubleEntityKey = e.Entity.Entity.Key
+            }.DoubleEntityKey.In(keyArrayAsCollection)).ToList());
+      }
+    }
+
+    private void PopulateEntities()
+    {
+      new MyEntity3 {
+        Entity = new MyEntity2 {
+          Entity = new MyEntity { Field = "ABC" }
+        }
+      };
+      new MyEntity3 {
+        Entity = new MyEntity2 {
+          Entity = new MyEntity { Field = "DEF" }
+        }
+      };
+      new MyEntity3 {
+        Entity = new MyEntity2 {
+          Entity = new MyEntity { Field = "GHI" }
+        }
+      };
+      new MyEntity3 {
+        Entity = new MyEntity2 {
+          Entity = new MyEntity { Field = "KLM" }
+        }
+      };
+      new MyEntity3 {
+        Entity = new MyEntity2 {
+          Entity = new MyEntity { Field = "NOP" }
+        }
+      };
+      new MyEntity3 {
+        Entity = new MyEntity2 {
+          Entity = new MyEntity { Field = "QRS" }
+        }
+      };
+
+      new ComplexKeyEntity(1, 1) { Name = "1 1" };
+      new ComplexKeyEntity(2, 2) { Name = "2 2" };
+      new ComplexKeyEntity(3, 3) { Name = "3 3" };
+      new ComplexKeyEntity(4, 4) { Name = "4 4" };
+      new ComplexKeyEntity(5, 5) { Name = "5 5" };
+      new ComplexKeyEntity(6, 6) { Name = "6 6" };
+      new ComplexKeyEntity(7, 7) { Name = "7 7" };
+      new ComplexKeyEntity(8, 8) { Name = "8 8" };
+    }
+
+    private void PopulateKeysCollections(Session session)
+    {
+      keyList = session.Query.All<MyEntity>().AsEnumerable().Take(3).Select(e => e.Key).ToList();
+      keyListAsEnumerable = new List<Key>(keyList);
+      keyListAsCollection = new List<Key>(keyList);
+      keyListAsInteface = new List<Key>(keyList);
+
+      keyArray = session.Query.All<MyEntity>().AsEnumerable().Take(3).Select(e => e.Key).ToArray();
+      keyArrayAsEnumerable = keyArray.AsEnumerable().ToArray();

```

Comment:
Why it is not enough to use just `AsEnumerable` call?

Reply:
It is not relevant since you encouraged me to rethink test collections one comment bellow.